Initial QEMU Support

colpane opened the pull request:
A **qemu-image** is added for booting IOT2050 on qemu system. You can test the not hardware dependent functionality using **qemu-image**.

All your comments/suggestions/criticisms are welcomed.
Thanks..

colpane commented:
@jan-kiszka @BaochengSu Please review and comment. :)

BaochengSu commented on Kconfig:
```diff
@@ -116,7 +116,11 @@ config SECURE_BOOT
+	  LUKS2 partitions.
+
+	  The reason of coupling the data encryption with the secure boot is to
+	  have a proper integrity chain rooted in hardware (OTP, RPMB, etc.)
```
This could be rebased into the first commit.

jan-kiszka replied:
First commit was wrong already, and fixing that will obsolete this one.

colpane replied:
There was a rebase issue, now it is aligned with master branch.

jan-kiszka commented on Kconfig:
```diff
@@ -30,6 +30,11 @@ config IMAGE_SWUPDATE
+config IMAGE_QEMU
+	bool "QEMU Image"
+	help
+	  QEMU image to boot image without real hardware.
```
Should mention the differences to the real device, e.g. the missing Arduino I/Os.

colpane replied:
```suggestion
	  QEMU image to boot image without real hardware. This image has
          missing or not working features especially dependent on hardware.
          Please consider limited functionality before using this (e.g. the
          missing Arduino I/Os).
```

Is this information enough/clear? What could I add more?

jan-kiszka replied:
What else is different? We only have one Ethernet port configured via the command line below. Only one UART, the debug one.

I would write it like this: "The device emulated by QEMU is not identical to a physical IOT2050. This leads to missing features such as ... Please be aware of potential problems when using the image for testing purposes."

colpane replied:
Done.

jan-kiszka commented on Kconfig:
```diff
@@ -110,11 +116,7 @@ config SECURE_BOOT
-
-	  The reason of coupling the data encryption with the secure boot is to
-	  have a proper integrity chain rooted in hardware (OTP, RPMB, etc.)
+	  Note: The persistent /var and /home partitions are not yet protected!
```
Unrelated changes.

colpane replied:
There was a rebase issue, now it is aligned with master branch.

jan-kiszka commented on conf/distro/iot2050-debian-qemu.conf:
```diff
@@ -0,0 +1,15 @@
+# Copyright (c) Siemens AG, 2023
+#
+# Authors:
+#  Enes �olpan <enes.colpan@siemens.com>
```
Is this UTF8 in your name and just a limitation of github in the visualization?

BaochengSu replied:
I guess this is a typo -- see recipes-core/images/iot2050-image-qemu.bb

colpane replied:
File is deleted, no problem anymore :)

jan-kiszka commented on conf/distro/iot2050-debian-qemu.conf:
```diff
@@ -0,0 +1,15 @@
+
+# used for SWUpdate boot
+ABROOTFS_IMAGE_RECIPE = "iot2050-image-qemu"
+VERITY_IMAGE_RECIPE = "iot2050-image-qemu"
```
Do we really need an own image, or can we also get away with just an own machine?

colpane replied:
Switched to machine instead of image.

jan-kiszka commented on recipes-core/images/iot2050-image-qemu.bb:
```diff
@@ -0,0 +1,42 @@
+IMAGE_INSTALL += "swupdate-complete-update-helper"
+IMAGE_INSTALL += "iot2050-watchdog"
+
+DTB_FILES = ""
```
This is really too much duplication and will be hard to maintain, constantly broken. Focus on the real differences.

jan-kiszka commented on recipes-kernel/linux/files/iot2050_defconfig_base:
```diff
@@ -691,3 +691,8 @@ CONFIG_SAMPLES=y
+CONFIG_SERIAL_AMBA_PL011_CONSOLE=y
+CONFIG_SCSI_VIRTIO=y
+CONFIG_DRM_VIRTIO_GPU=y
+CONFIG_VIRTIO_INPUT=y
```
Missing newline at end of file.

And this defconfig extension should probably go into the extra set as it is not need for the IOT2050 base.

jan-kiszka replied:
..and that as an own commit.

colpane replied:
Done.

jan-kiszka commented on wic/iot2050-qemu.wks.in:
```diff
@@ -0,0 +1,20 @@
+part /var --source rootfs --rootfs-dir=${IMAGE_ROOTFS}/var --label var --fstype=ext4 --mkfs-extraopts "-T default" --ondisk sda --align 1024  --size 512M
+part /home --source rootfs --rootfs-dir=${IMAGE_ROOTFS}/home --label home --fstype=ext4 --mkfs-extraopts "-T default" --ondisk sda --align 1024  --size 512M
+
+bootloader --ptable gpt --append "rootwait ${EXTRA_KERNEL_PARAMS}"
```
Again, too much duplication. Why do we need an own image layout?

colpane replied:
Removed file and used swu.wks file with a variable.

colpane commented on conf/machine/iot2050-qemu.conf:
```diff
@@ -0,0 +1,18 @@
+
+CONSOLE_KERNEL_PARAMS = ""
+
+PREFERRED_PROVIDER_u-boot-${MACHINE} = "u-boot-qemu-arm64"
+PREFERRED_PROVIDER_u-boot-${MACHINE}-config = "u-boot-qemu-arm64"
```
Somehow this generates u-boot-pg1 instead of qemu. Will fix it.

jan-kiszka replied:
Hmm, where should that qemu u-boot come from? A Debian package?

colpane replied:
It comes from isar-cip-core repository.
https://gitlab.com/cip-project/cip-core/isar-cip-core/-/tree/master/recipes-bsp/u-boot

jan-kiszka replied:
Ah, of course!

colpane replied:
Fixed.

jan-kiszka commented on README.md:
````diff
@@ -58,6 +58,28 @@ The SDK is also available as docker image. To import it into a docker host, run
+qemu-system-aarch64 is installed.
+
+```shell
+qemu-system-aarch64 -bios ./firmware.bin -m 2G -netdev user,id=net,hostfwd=tcp:127.0.0.1:22222-:22
````
I suspect there is ./firmware.bin after building the QEMU image, is there? Better reference the one in the deploy folder.

colpane replied:
It will be fixed with upper u-boot bug. Then I will reference the exact path.

colpane commented on Kconfig:
```diff
@@ -58,10 +63,11 @@ config KAS_INCLUDE_MAIN
 	default "kas-iot2050-swupdate.yml" if IMAGE_SWUPDATE
 	default "kas-iot2050-boot-pg1.yml" if IMAGE_BOOT_PG1
 	default "kas-iot2050-boot-pg2.yml" if IMAGE_BOOT_PG2
+	default "kas-iot2050-qemu.yml" if IMAGE_QEMU
```
Swupdate support is already included in qemu. The reference for qemu was the image with swupdate support. Should we have two different qemu image for swupdate image and w/o swupdate example image?

jan-kiszka replied:
I see now. I would suggest to permit both image variants. Could be achieved by making the qemu yaml an option file that just switches the machine.

colpane replied:
Yes that would be definitely better. I will check and change the related parts.

jan-kiszka commented on Kconfig:
```diff
@@ -95,7 +95,7 @@ config KAS_INCLUDE_LXDE
 	default "kas/opt/lxde.yml"
 	depends on LXDE
-
+     
```
Stray whitespace chance.

jan-kiszka commented on recipes-core/images/iot2050-image-base.bb:
```diff
@@ -12,7 +12,7 @@ inherit image
 DESCRIPTION = "IOT2050 Debian Base Image"
 
-IMAGE_INSTALL += "u-boot-iot2050-config"
+IMAGE_INSTALL += "${@('u-boot-' + d.getVar('MACHINE')) if d.getVar('QEMU_IMAGE') == '1' else 'u-boot-${MACHINE}-config' }"
```
u-boot-*-config is needed for interacting with the U-Boot env. I don't think we have a persistent env with U-Boot under QEMU. In any case, we don't need u-boot-<machine> in that case.

colpane replied:
A new yml file is created for compiling u-boot qemu version. Removed from image.

jan-kiszka commented on recipes-core/images/iot2050-image-swu-example.bb:
```diff
@@ -40,6 +40,6 @@ IMAGE_INSTALL += "efibootguard libebgenv0"
 IMAGE_INSTALL += "swupdate-handler-roundrobin"
 IMAGE_INSTALL += "swupdate-complete-update-helper"
-IMAGE_INSTALL += "iot2050-watchdog"
+IMAGE_INSTALL += "${@ '' if d.getVar('QEMU_IMAGE') == '1' else 'iot2050-watchdog' }"
```
Will things break if that is left in? But, yeah, we are missing some watchdog, see also https://gitlab.com/cip-project/cip-core/isar-cip-core/-/issues/66

colpane replied:
I think it will not be a problem. However, I thought that removing the package for qemu would make no watchdog support feature more visible. What do you suggest at the end?

jan-kiszka replied:
Change as little as needed to get a system starting that shows no errors.

colpane replied:
recipes-app/board-conf-tools/board-conf-tools_0.1.bb
This package could be removed as it gives error at the end of start process. However, it does not ruin booting process.

jan-kiszka replied:
Right, but that would be a good example for removing something that cannot work (given the existing device model in QEMU). Here, I would vote for removing, also to avoid confusion.

colpane replied:
Done.

jan-kiszka commented on Kconfig:
```diff
@@ -50,6 +50,14 @@ config IMAGE_BOOT_PG2
+
+	  WARNING: Do not flash this image onto your device. Otherwise you
+	  risk to BRICK THE IOT2050!
+
```
Nope, please not. Rather model that similar to isar-cip-core where U-Boot forms a helper for the QEMU start script. Maybe even consider a start script over lengthy README.

colpane replied:
Sure, I will write a start-qemu script. This is for generating firmware.bin file for qemu, so we will keep this image type right?

jan-kiszka replied:
https://gitlab.com/cip-project/cip-core/isar-cip-core/-/blob/master/kas/opt/ebg-swu.yml#L33 - this is how the u-boot firmware image is hooked up in isar-cip-core.

colpane replied:
If we switch to this model, then we have to remove iot2050-watchdog. Otherwise, there is a conflict for compiling u-boot for qemu.

colpane replied:
Done, switched that model and removed iot2050-watchdog.

jan-kiszka commented on start-qemu-iot2050.sh:
```diff
@@ -0,0 +1,82 @@
+		-initrd ${INITRD_FILE} \
+		${QEMU_COMMON_OPTIONS} "$@"
+fi
+
```
Extra new-line at EOF.

jan-kiszka commented:
Looks really good, just two style details! Will also give it a try soon, but that should not delay things.

colpane commented:
Thank you Jan, especially for your valuable comments/suggestions. Will check last two style things.

jan-kiszka commented:
When addressing remarks on patches, fold the changes into the affected patches, don't stack them on top.

colpane commented:
> When addressing remarks on patches, fold the changes into the affected patches, don't stack them on top.

Like having separate commits for each goal?

jan-kiszka commented:
That is the foundation of a patch series in any case. And if you modify something in v2 or later, only add patches if you address new separate topics.

jan-kiszka commented:
Quick test feedback: Normal image works, but the SWUpdate / UEFI variant seems to get stuck past the kernel's EFI stub (no output on virtual UART). Does that work for you?

colpane commented:
> Quick test feedback: Normal image works, but the SWUpdate / UEFI variant seems to get stuck past the kernel's EFI stub (no output on virtual UART). Does that work for you?

Yes, both of them were fine for me.

jan-kiszka commented:
Yes, I'm using the script.

Hmm, now it works: Switched to the normal image, then back to the swupdate one, and now it's fine. Need to perform a clean build into the latter tomorrow to really rule out dependency issues.

jan-kiszka commented:
All still good after a cleansstate rebuild. Must have been a partial rebuild fallout.

jan-kiszka commented:
But now I'm testing secure boot + encryption. I'm not seeing TPM support in your start script, though... Maybe that feature should be grayed out for now?

colpane commented:
You are right, I should gray out it if IMAGE_QEMU is selected. Also, I haven't test secure boot + encryption supported image.

jan-kiszka commented on Kconfig:
```diff
@@ -148,11 +148,29 @@ config RPMB_SETUP
 	string
 	default "kas/opt/secure-boot.yml"
-	depends on SECURE_BOOT
+	depends on SECURE_BOOT && !IMAGE_QEMU
```
This is redundant as `SECURE_BOOT` has this precondition already.

colpane replied:
Done.

jan-kiszka reviewed (changes requested):
Looks good to me otherwise.

colpane commented:
Comments/problems are fixed and squashed unnecessary commits. Ready to merge 👍
@jan-kiszka @BaochengSu

jan-kiszka commented on wic/iot2050-swu.wks.in:
```diff
@@ -17,4 +17,4 @@ part --source empty --align 1024 --fixed-size 4G --uuid "${ABROOTFS_PART_UUID_B}
 part /home --source rootfs --rootfs-dir=${IMAGE_ROOTFS}/home --label home --fstype=ext4 --mkfs-extraopts "-T default" --ondisk sda --align 1024  --size 512M
 
-bootloader --ptable gpt --append "console=ttyS3,115200n8 earlycon=ns16550a,mmio32,0x02810000 rootwait watchdog.handle_boot_enabled=0 ${EXTRA_KERNEL_PARAMS}"
+bootloader --ptable gpt --append "${CONSOLE_KERNEL_PARAMS} rootwait watchdog.handle_boot_enabled=0 ${EXTRA_KERNEL_PARAMS}"
```
Is that a "no new line at end of file"?

BaochengSu replied:
@colpane , If this is fixed then we can merge it :)

colpane replied:
I can fix it, however all the wks.in files do not have a new line at the end of file and this wks.in file also did not have a new line at the end of file before I changed it :) Do I still need to add a new line?

jan-kiszka replied:
The affected file currently does have a newline at its end.

colpane replied:
Correct, it does not have, but other wks.in files (https://github.com/siemens/meta-iot2050/blob/master/wic/iot2050-swu-secure.wks.in) do not have it too. Why only this file should have the new line?

jan-kiszka replied:
Just don't remove the newline removal from your patch and resubmit. Avoid style regressions.

colpane replied:
Ah, sorry now I see the icon :) Thanks, fixed.

jan-kiszka reviewed (approved):
Just a style nit, looks good.
